packages/react-router-dom/modules/BrowserPrompt.js

```diff
@@ -0,0 +1,80 @@
+import React, { PropTypes } from 'react'
+import { Prompt } from 'react-router'
+
+const BEFOREUNLOAD = "beforeunload"; // see: http://stackoverflow.com/questions/39094138/reactjs-event-listener-beforeunload-added-but-not-removed
+
+var UnloadComponent = ComposedComponent => class extends React.Component {
+  static contextTypes = {
+    router: PropTypes.shape({
+        history: PropTypes.shape({
+            block: PropTypes.func.isRequired
+        }).isRequired
+    }).isRequired
+  }
+
+  static propTypes = {
+    when: PropTypes.bool,
+    message: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.string
+    ]).isRequired,
+    beforeUnload: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.bool
+    ])
+  }
+
+  constructor(props) {
+    super(props)
+  }
+
+  onbeforeunload(e) {
+    var dialogText = "Changes you made may not be saved."
+    e.returnValue = dialogText
+    return dialogText
+  }
+
+  enableUnload(_onbeforeunload) {
```

Comment:
Agreed, we used them to indicate a serious part of the (private) code that our juniors are not allowed to touch 😄